aptos-move/move-examples/large_packages/sources/large_packages.move

```diff
@@ -1,30 +1,65 @@
-/// This provides a framework for uploading large packages. In each pass, the caller pushes more
-/// code by calling `stage_code`. In the last call, the caller can set the optoinal `publish` and
-/// the package will be published inline, saving an extra transaction and additional storage costs.
-/// Currently this module does not support modules that are larger than 63KB as that is the maximum
-/// that can fit within a transaction and this framework does not split up individual modules.
+/// This provides a framework for uploading large packages to standard accounts or objects.
+/// In each pass, the caller pushes more code by calling `stage_code_chunk`.
+/// In the final call, the caller can optionally set `publish_to_account`, `publish_to_object`, or `upgrade_object_code`.
+/// If any of these options are set, the package will be published inline, saving an extra transaction and additional storage costs.
 module large_packages::large_packages {
     use std::error;
+    use std::option::{Self, Option};
     use std::signer;
     use std::vector;
+    use aptos_std::smart_table::{Self, SmartTable};
 
-    use aptos_framework::code;
+    use aptos_framework::code::{Self, PackageRegistry};
+    use aptos_framework::object::{Object};
+    use aptos_framework::object_code_deployment;
 
     /// code_indices and code_chunks should be the same length.
     const ECODE_MISMATCH: u64 = 1;
+    /// The publishing flags should either all be false, or only one should be true.
+    const EINVALID_PUBLISHING_FLAGS: u64 = 2;
+    /// Object reference should be provided when upgrading object code.
+    const EMISSING_OBJECT_REFERENCE: u64 = 3;
 
-    struct StagingArea has drop, key {
+    struct StagingArea has key {
         metadata_serialized: vector<u8>,
-        code: vector<vector<u8>>,
+        code: SmartTable<u64, vector<u8>>,
+        last_module_idx: u64,
     }
 
-    public entry fun stage_code(
+    public entry fun stage_code_chunk(
         owner: &signer,
-        metadata_serialized: vector<u8>,
+        metadata_chunk: vector<u8>,
         code_indices: vector<u16>,
         code_chunks: vector<vector<u8>>,
-        publish: bool,
+        publish_to_account: bool,
+        publish_to_object: bool,
+        upgrade_object_code: bool,
```

Comment:
Would it be better to have the object publish/upgrade command in a separate API? i.e. `stage_object_code_chunk`?